### ISSUE TITLE
Router support for handler()s that are async functions 

### ISSUE THREAD

**Issue by jeffposnick** [GoogleChrome/sw-helpers#200]
**Library Affected**:
*sw-routing*

As @surma points out at https://github.com/w3c/ServiceWorker/issues/1068#issuecomment-278368743, there's a desire to start using `async` functions as the input to service worker event handlers' `waitUntil()` or `respondWith()`.

Currently, the `handler` that's passed to our `Route`s needs to be an object which exposes a `handle` method, and that method needs to return a `Promise<Response>`.

@surma pointed out the use case of supplying an `async` function inside of an `install` event handler's `waitUntil()`, and I'm not sure that it makes sense to expose anything for that inside of `sw-helpers`, since we expect folks who use the libraries for precaching to use the `install` event handler that `sw-lib` manages. But I could theoretically see the advantage of opening up `Route`s used for runtime strategies, so that they're more flexible and can take in `async` functions as their `handler`s.

Not 100% sold on this idea, but it could be worth prototyping and seeing if it make for a more fluent interface.

**Comment by jeffposnick** [comment]
It's worth noting that all the handlers we ship as part of `sw-runtime-caching` are specified in the source as `async` functions, but they're transpiled by Babel to use generators during the build process.

E.g. https://github.com/GoogleChrome/sw-helpers/blob/c26d5b81f1ec4c8417d4e657712e3518d3e43826/packages/sw-runtime-caching/src/lib/stale-while-revalidate.js#L51

**Comment by gauntface** [comment]
@jeffposnick Can you explain the problem?

I'm not entirely sure why event.waitUntil(async function() {}()) is a problem and I'm not 100% sure what this bug is requesting / about.

**Comment by surma** [comment]
Sorry for letting this go quiet, got distracted and forgot about it.

(The original is [here](https://github.com/w3c/ServiceWorker/issues/1068) ).
**TL;DR:** If you forget to actually invoke the async IIFE, nothing will be executed but not error will be thrown either. It’s a pain to debug why your SW stopped working.

Would be interested to hear what you think, @gauntface. Will get back on top of this.

**Comment by gauntface** [comment]
I still feel like I'm missing something.

This isn't an issue with the language or with the API, just if the developer writes some code that doesn't act as intended (i.e. they forget to execute their function) - it has nothing to do with async or async support - is that right?

**Comment by surma** [comment]
In a way it’s an issue with the API as accepts any kind of value for `waitUntil` when it should be really awaiting something that is a promise. As mentioned in the original thread we _should_ have required to settle with a specific value so that we can catch if developers are doing something stupid. I am investigating if we can at least issue a warning.

**Comment by gauntface** [comment]
So the problem is a spec issue.

event.waitUntil() should force a promise to be returned and throw on anything else.

Whats the issue with sw-helpers? That if a developer uses this library without transpilation we are going to cause them issues?
